airbyte-cdk/java/airbyte-cdk/core/src/main/kotlin/io/airbyte/cdk/integrations/util/ConnectorExceptionHandler.kt

```diff
@@ -0,0 +1,175 @@
+import io.airbyte.commons.exceptions.ConfigErrorException
+import io.airbyte.commons.exceptions.ConnectionErrorException
+import io.airbyte.commons.exceptions.TransientErrorException
+import io.airbyte.protocol.models.v0.AirbyteConnectionStatus
+import io.airbyte.protocol.models.v0.AirbyteMessage
+import io.github.oshai.kotlinlogging.KotlinLogging
+import java.util.function.Consumer
+import kotlin.system.exitProcess
+import org.jetbrains.annotations.VisibleForTesting
+
+private val LOGGER = KotlinLogging.logger {}
+
+const val COMMON_EXCEPTION_MESSAGE_TEMPLATE: String =
+    "Could not connect with provided configuration. Error: %s"
+
+const val DATABASE_CONNECTION_ERROR: String =
+    "Encountered an error while connecting to the database error"
+
+enum class FailureType {
+    CONFIG,
+    TRANSIENT
+}
+
+data class ConnectorErrorProfile(
+    var errorClass: String,
+    var regexMatchingPattern: String,
+    var failureType: FailureType,
+    var externalMessage: String,
+    var sampleInternalMessage: String,
+    var referenceLink: List<String>
+)
+
+/**
+ * This abstract class defines interfaces that will be implemented by individual connectors for
+ * translating internal exception error messages to external user-friendly error messages.
+ */
+open class ConnectorExceptionHandler {
+    @kotlin.jvm.JvmField
+    var DATABASE_READ_ERROR: String = "Encountered an error while reading the database"
+
+    protected open var connectorErrorDictionary: MutableList<ConnectorErrorProfile> =
+        mutableListOf()
+
+    /**
+     * Handles exceptions thrown by the connector. This method is the main entrance for handling
+     * exceptions thrown by the connector. It checks if the exception is a known exception, and if
+     * so, it emits the appropriate trace and external user-friendly error message. If the exception
+     * is not known, it rethrows the exception, which becomes a system error.
+     */
+    fun handleException(
+        e: Throwable,
+        cmd: Command,
+        outputRecordCollector: Consumer<AirbyteMessage>
+    ) {
+        ApmTraceUtils.addExceptionToTrace(e)
+        val rootException: Throwable? = getRootException(e)
+        val externalMessage: String? = getExternalMessage(rootException)
+        // error messages generated during check() needs special handling
+        if (cmd == Command.CHECK) {
+            outputRecordCollector.accept(
+                AirbyteMessage()
+                    .withType(AirbyteMessage.Type.CONNECTION_STATUS)
+                    .withConnectionStatus(
+                        AirbyteConnectionStatus()
+                            .withStatus(AirbyteConnectionStatus.Status.FAILED)
+                            .withMessage(externalMessage),
+                    ),
+            )
+        } else {
+            if (checkErrorType(rootException, FailureType.CONFIG)) {
+                AirbyteTraceMessageUtility.emitConfigErrorTrace(e, externalMessage)
```

Comment:
we might want to `log.error(e) { ... }` here before exiting so that the exception stack trace gets printed into the logs

Reply:
this is another part that I'm not very familiar with, so would the code above:

```
AirbyteTraceMessageUtility.emitConfigErrorTrace(e, externalMessage)
```
emit the trace and external message into the log?

Reply:
No, `AirbyteTraceMessageUtility.emitConfigErrorTrace` only emits a TRACE airbyte message, it doesn't populate the connector log. Both are needed.

Reply:
I see! It seems we didn't do this in the previous implementation either. I will add this.

Reply:
fixed.